Custom factual correctness metric

exfalsoquodlibet opened the pull request:
This PR introduces the first of three custom evaluation metrics for assessing the quality of LLM-generated answers against an ideal ("ground-truth") answer: Factual Correctness.

The metric uses an LLM-as-a-Judge approach—an LLM is prompted to evaluate factual overlap between the generated and ground-truth answers.

Our implementation of Factual Correctness is inspired by the [Ragas framework](https://docs.ragas.io/en/v0.1.21/concepts/metrics/answer_correctness.html) and measures a form of factual precision:
How many of the key facts in the generated answer can be found or inferred from the ground-truth?

### Implementation steps:

1. The LLM identifies key factual statements in both the generated and ideal answers.

2. The LLM [then classifies these into Truth Positives (TP) and False Positives (FP)](https://github.com/alphagov/govuk-chat-evaluation-prototype/blob/fcacc10cf9baad59627cb0092881cb27fdb81b68/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/template.py#L4-L24)
This is implemented in [`_classify_statements`](https://github.com/alphagov/govuk-chat-evaluation-prototype/blob/fcacc10cf9baad59627cb0092881cb27fdb81b68/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py#L131)

3. [Factual Correctness is computed as Precision: `|TP| /  |TP|  + |FP|`, a score in [0,1]](https://github.com/alphagov/govuk-chat-evaluation-prototype/blob/fcacc10cf9baad59627cb0092881cb27fdb81b68/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py#L155).

5. The classified facts ("confusion metrics") are returned as `reason` to make it compatible with DeepEval further logic and pipelines, [here](https://github.com/alphagov/govuk-chat-evaluation-prototype/blob/fcacc10cf9baad59627cb0092881cb27fdb81b68/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py#L86-L101).

For the implementation:
- I followed [DeepEval's custom metric guidelines](https://www.deepeval.com/docs/metrics-custom#rules-to-follow-when-creating-a-custom-metric) 
- and used [one of deepeval's own metric implementation as a reference](https://github.com/confident-ai/deepeval/blob/main/deepeval/metrics/hallucination/hallucination.py) )


### Some other considerations:

- `FactualCorrectnessMetric` is a subclass of deepeval `BaseMetric`. Working within this base class introduced some type hinting limitations due to its generality. While this led to occasional friction, I worked within its constraints to preserve compatibility. 
- The metric behaves differently when using GPTModel vs. custom models or other models, because DeepEval returns (output, cost) tuples only for GPTModel, while other models return strings. This required additional handling for cost-aware evaluation.
- `measure` method delegates to `a_measure()` when `async_mode` is True. This is to allows the `async_mode` set by user to route to the async or sync implementation.

### Tests

I added two types of tests:
- Unit tests: to validate internal logic and structure.
- Metric behavioural tests: to verify that the metric yields correct factuality scores. These require live OpenAI API access and are skipped by default to avoid incurring API costs. To run them: `RUN_OPENAI_TESTS=1 uv run pytest`

exfalsoquodlibet commented:
As a side note, I would like to propose we rename this metric as `factual_precision` to:
- better reflects what the metric is actually measuring
- help manage stakeholders' expectations re: what we can auto-evaluate vs manually evaluate

kevindew reviewed (commented):
Posting partial review at this point - looks great, but know I'm going to take a while so pushing some things through now

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+        ] = FactualCorrectnessTemplate,
+    ):
+        self.using_gpt_model = isinstance(model, GPTModel)
+        self.model = model
```
Thanks for this Kev, good shout. 

I tested it and it works to out case, returning

```
>>> from deepeval.models.llms.openai_model import GPTModel
>>> my_model = GPTModel(model="gpt-4o-mini", temperature=0.0)
>>> initialize_model(my_model)
(<deepeval.models.gpt_model.GPTModel object at 0x17a47bc70>, True)
```

and with the custom Amazon Nova model (not yet ported to this repo)
```
>>>from ..custom_judge_models.aws_bedrock_nova_models import (
    CustomAmazonNovaJudge,
)
>>>nova_model = CustomAmazonNovaJudge()
>>> initialize_model(nova_model)
(<answer_generation.src.custom_judge_models.aws_bedrock_nova_models.CustomAmazonNovaJudge object at 0x17a4c3dc0>, False)
```


I was using a false logic before to check "nativeness", 
```
if isinstance(model, DeepEvalBaseLLM):
            self.using_native_model = False
            self.model = model
        else:
            self.using_native_model = True
            self.model = GPTModel(model=model)
```

based on an old version of deepeval, that misleadingly returned `False` when passing 
```
my_model = GPTModel(model="gpt-4o-mini", temperature=0.0)
isinstance(model, DeepEvalBaseLLM)
```

I'll implement this change after you have finalised your review, unless you are happy for me to make these changes now.

kevindew replied:
Happy for you to make them now - will take me a while still to complete the review.

exfalsoquodlibet replied:
Done in: https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/58c3bb9e8dd632fc7d731125924e0f6b0bc13c0a

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+        self.async_mode = async_mode
+        self.strict_mode = strict_mode
+        self.evaluation_template = evaluation_template
+        self.evaluation_cost = 0 if self.using_gpt_model else float("nan")
```
I had issue with `Pylance` complaining about type-checking error because Python does not allow arithmetic between a (potential) `None` and numeric types, and `evaluation_cost` is defined as as an optional float with a default value of `None` in `baseMetric`: https://github.com/confident-ai/deepeval/blob/5798fccae9cf75e933ce4009ac7d2b28e9547a0f/deepeval/metrics/base_metric.py#L26

Simple reproducible example:

```
a: float = 5.0
b: float | None = None

result = a + b 
```

complains that:
```
Operator "+" not supported for types "float" and "None"
```

As an alternative work-around, I could do:

```
result = a + (b or 0.0)
```

So in our code:

Instead of [line 45](https://github.com/alphagov/govuk-chat-evaluation-prototype/blob/fcacc10cf9baad59627cb0092881cb27fdb81b68/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py#L45), revert back to as in deepeval code:
```
self.evaluation_cost = 0 if self.using_native_model else None
```

and then, at [line 135](https://github.com/confident-ai/deepeval/blob/5798fccae9cf75e933ce4009ac7d2b28e9547a0f/deepeval/metrics/answer_relevancy/answer_relevancy.py#L135):
```
self.evaluation_cost = (self.evaluation_cost or 0.0) + cost
```
instead of
```
self.evaluation_cost += cost
```

exfalsoquodlibet replied:
Done in https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/a5a22fef5f9438527c1beccc7249fc183d8c4d78

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+            else:
+                logging.info(
+                    "Async mode is disabled. Running the synchronous function."
+                )
```
Good point. I'll remove the logging, and also agree that we are likely not to use the synchronous implementation so we can get rid of that and raise a `NotImplementedError`.

exfalsoquodlibet replied:
Removed in https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/3decfe6d9eecec8f5f68215e0475b6f94cd5f732

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+                self.confusion_matrix = self._classify_statements(
+                    test_case.actual_output, test_case.expected_output or ""
+                )
+                logging.info(f"Confusion matrix: {self.confusion_matrix}")
```
Yes, it was very useful for debugging. What's the best way to achieve this? Should I leave this in and just commented out?

kevindew replied:
we could put it as logger.debug ? I'm not sure what we've got set-up for running logger in debug mode but this seems the sort of detail.

exfalsoquodlibet replied:
I tried to achieve this in this way: https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/8251242cc8f0b7a73e8779fa0a07698fdbd1064c

but it was not a good idea, I realise now, since we get all kind of debug statements printed into the terminal now... so I'll revert back to stream_handler log level INFO and perhaps, we could:

- handle this with a `print` statement instead
- add a filter,  `console_handler.addFilter(SpecificDebugFilter())` ref: https://stackoverflow.com/a/879937/4014292

Happy to get your steer on this.

kevindew replied:
Sure. So my expectations were that we'd set this to logger.debug but not change the default debugging level.

Then only when we actually wanted to debug the confusion matrices would we temporarily adjust the debugging level - expecting that it would be noisy.

That's based on my expectation that debugging the confusion matrices should be an exceptional action and not one that we want to do each time we run.

exfalsoquodlibet replied:
I see, thanks. I removed the commit with the change to the stream_handler log level setting.

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+                logging.error("Failed to parse fallback JSON.", exc_info=inner_e)
+                return ClassifiedFacts(TP=[], FP=[], FN=[])
+
+    def _classify_statements(
```
Agree that we are likely not to use the synchronous implementation so we can get rid of that and raise a `NotImplementedError` as you suggest in your previous comment.
I'm trying to think if and when we may need the sync version, but it is mostly for local testing.

exfalsoquodlibet replied:
Refactored to only include the async methods in https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/d77c9eaed90d97f240dbce719cbf9cdfb8ad3cff

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+        check_llm_test_case_params(test_case, self._required_params, self)
+        logging.info(
+            "Async mode is enabled. Using event loop to run the async function."
+        )
```
Removed

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+            self.confusion_matrix = await self._a_classify_statements(
+                test_case.actual_output, test_case.expected_output or ""
+            )
+            logging.info(f"Confusion matrix: {self.confusion_matrix}")
```
See comment above https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39#discussion_r2088347200

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+                prompt, schema=FactClassificationResult
+            )
+            if isinstance(cost, (int, float)):
+                self.evaluation_cost = (self.evaluation_cost or 0.0) + float(cost)
```
Yes, done

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,181 @@
+        """Synchronously evaluate the factual correctness of a test case."""
+        check_llm_test_case_params(test_case, self._required_params, self)
+
+        with metric_progress_indicator(self):
```
I think this is now superseded since we no longer have a a synchronous implementation and the async one does have a `_show_indicator` argument https://github.com/alphagov/govuk-chat-evaluation-prototype/blob/d77c9eaed90d97f240dbce719cbf9cdfb8ad3cff/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py#L61-L63

kevindew commented on tests/conftest.py:
I imagine it'd be good to use load_dotenv in this file so it can load the OpenAI key from your .env file

exfalsoquodlibet replied:
Done now that we are importing it via conftest.py if required https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/e911acf767934045373919f57fab4c33f20c57e4

kevindew reviewed (commented):
Amazing work @exfalsoquodlibet, I've finally got around to finishing my review. Sorry it took so long. 

I think the implementation looks good, with perhaps a couple of areas of simplification. I have made a pretty meaty suggestion on test structure and how we'd test with real OpenAI so very happy to chat those over with you.

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,141 @@
+        self.strict_mode = strict_mode
+        self.evaluation_template = evaluation_template
+        self.evaluation_cost = 0 if self.using_native_model else None
+        self.confusion_matrix: Optional[ClassifiedFacts] = None
```
I’m not opposed to the change, but I want to clarify the reasoning:

Defining `confusion_matrix` as a class attribute is safe here because the entire ClassifiedFacts object is always reassigned, never mutated in-place, preventing unintended shared state.

However, the class attribute is effectively only a default (None) that gets shadowed by an instance attribute once `self.confusion_matrix = ...` is assigned.

This means in practice, each instance has its own `confusion_matrix` after initialization or use, so no state is shared between instances.

Clarity of intent (it makes it immediately obvious that each instance should have its own independent confusion_matrix) and idiomatic Python, it is probably safer to define it  as an instance attribute , making the instance-specific nature explicit.

Happy to discuss further!

exfalsoquodlibet replied:
I have thought move `async_mode` to a class attribute as I think, reasoning about the above, it makes more sense: https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/95d528a9887db960d578de537f95eaa9f7b35b2f

exfalsoquodlibet replied:
This is interesting. This is my understanding (**happy to be corrected**!).

```
class Example:
    value: Optional[int] = None
```

In Python, this creates a class variable, not an instance variable. That means value is shared across all instances unless it's explicitly overridden (e.g., in `__init__` or by assigning to `self.value` at runtime). It’s not the standard way to define default instance attributes.

For deepeval's

```
class BaseMetric:
    success: Optional[bool] = None
    reason: Optional[str] = None
    ...
```

I think their approach is a bit peculiar. These are still class attributes IMHO. When a subclass of BaseMetric is instantiated, those values:

- act as defaults, but
- are expected to be overwritten by defining them at the class level in the subclass ot setting them in their __init__ method or at runtime (e.g. self.success = ... in the metric logic).

Until that happens, success and reason remain shared across instances.

In short: they behave like instance attributes once assigned, but they start life as shared class attributes, which can be confusing and potentially error-prone if the values were mutable.

kevindew replied:
Thanks, for explaining yeah I've found myself a bit confused by it as the class attributes side seems to conflict with the instance variable type annotations - I guess it's just a consequence of typing added late to the language and a needing to be a bit bodged. 

You are indeed right that they're shared until set:

```
>>> from typing import Optional
>>> class Example:
...     a: Optional[int] = None
...
>>> instance1 = Example()
>>> instance1.a = 2
>>> instance2 = Example()
>>> type(instance2.a)
<class 'NoneType'>
>>> type(Example.a)
<class 'NoneType'>
>>> Example.a = 1
>>> type(Example.a)
<class 'int'>
>>> instance2.a
1
>>> instance1.a
2
```

Feel free to just resolve this comment without making a change.

exfalsoquodlibet replied:
@kevindew and I discussed offline. We'll keep the original instance variable implementation.

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
Really good point. In https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/5cd7665521d378f25b70ba07de6b9217efd72122

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,141 @@
+            self.confusion_matrix = await self._a_classify_statements(
+                test_case.actual_output, test_case.expected_output or ""
+            )
+            logging.debug(f"Confusion matrix: {self.confusion_matrix}")
```
Yes, good point, sorry that was slopping thinking from my side: https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/095af4020c6c26322c00ed9cf876f2566acf72de

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,141 @@
+                data_model = FactClassificationResult(**data)
+                return data_model.classified_facts
+            except Exception as inner_e:
+                logging.error("Failed to parse fallback JSON.", exc_info=inner_e)
```
In https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/35514d00b30b5f9d5f2adf05d20bafd5be6a29a7

exfalsoquodlibet replied:
Good suggestion. I updated the `_a_classify_statements` method to be passed an `input` as well to allow for this extra logging. In https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/44e9ded4a42d65c86b73eccec2d640f8660699ed

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/data_models.py:
```diff
@@ -12,6 +12,9 @@
 )
 from deepeval.models.llms.openai_model import GPTModel
 
+from .custom_deepeval.metrics.factual_correctness.factual_correctness import (
```
In https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/35514d00b30b5f9d5f2adf05d20bafd5be6a29a7

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,141 @@
+        strict_mode: bool = False,
+        evaluation_template: Type[
+            FactualCorrectnessTemplate
+        ] = FactualCorrectnessTemplate,
```
yes, good suggestion. In https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/45434976013e86b9420fab065753b444ecb9d783
(some of the changed tests will be further edited via other commit as I address your test-related comments)

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+        classified_facts=ClassifiedFacts(TP=["fact1"], FP=["fact2"], FN=[])
+    )
+
+
```
Thanks a lot for this comment, Kev. I hope I have addressed this (and the other comments on mocked unit tests) in https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/c0891aac6cc9811c4a554db63a46eeb5c5d9f2bc

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+    ):
+        mock_native_model.a_generate = AsyncMock(
+            return_value=(Mock(classified_facts="mocked_facts"), 0.01)
+        )
```
Addressed as part of https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/c0891aac6cc9811c4a554db63a46eeb5c5d9f2bc

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+            model=mock_native_model, threshold=0.7, include_reason=True
+        )
+
+        metric._a_classify_statements = AsyncMock(return_value=sample_classified_facts)
```
Addressed as part of https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/c0891aac6cc9811c4a554db63a46eeb5c5d9f2bc

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+    @pytest.mark.asyncio
+    @patch(
+        "govuk_chat_evaluation.rag_answers.custom_deepeval.metrics.factual_correctness.factual_correctness.trimAndLoadJson"
+    )
```
Addressed as part of https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/c0891aac6cc9811c4a554db63a46eeb5c5d9f2bc

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+        assert metric.is_successful() is False
+
+
+# -------- DeepEval metrics tests --------
```
Really good suggestion, thanks. In https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/45fb76c1249d087e453aba95e1f5798bc2319399

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+        model=GPTModel(model="gpt-4o", temperature=0),
+        include_reason=False,
+    )
+    assert_test(test_case, [metric])
```
I have also found it cryptic. Reviewing the code of the method, it looks indeed like an integration tests. I followed what Deepeval does for all the other metrics, so included it as I thought it was standard approach for the library.

It basically runs a full metric evaluation on a test case using a real model and checks if the metric passes their thresholds. If the metric fails, it raises an `AssertionError` with detailed failure information.
https://github.com/confident-ai/deepeval/blob/ee94734e0da738044db63333b65ee5264500ab72/deepeval/evaluate/evaluate.py#L57

kevindew replied:
Seems very reasonable to include it - we can always remove it if it's crypticness causes problems (especially if fails are hard to debug)

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+    if not api_key:
+        raise RuntimeError(
+            "RUN_OPENAI_TESTS is set, but OPENAI_API_KEY is not defined."
+        )
```
More elegant solution - thanks. Done in https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/dfe82ccb8ee96cdb57e1f71a615b5fa197178fca with also https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/e911acf767934045373919f57fab4c33f20c57e4

exfalsoquodlibet commented on tests/rag_answers/test_deepeval_custom_metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,482 @@
+)
+
+
+@pytest.mark.skipif("not run_openai_tests", reason="openai is expensive")
```
[question] Thanks Kev, this looks great. I have done the marker change, but I have a question re: test github action before implementing it. I assume we do not want to execute this tests by CI every time we commit to a PR. 

Or is that your preferred approach?

exfalsoquodlibet replied:
Done in https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/c500804a7811329b70eb3df71a6c8af23406e243

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            result = await metric.a_measure(test_case)
+
+            # expected score based on real implementation
+            expected_score = metric._calculate_score()
```
We should use a hardcoded number here rather than asserting based off the private method. Reasons to do so are: it's easier to reason with an actual number, it makes it clearer that it actually returns an expected type, as the code itself is already using the private method it conceals issues with that method.

exfalsoquodlibet replied:
Done as part of https://github.com/alphagov/govuk-chat-evaluation-prototype/pull/39/commits/f16d1789fc5e0c158f51aad2d711c4071af934af (also parametrised to test the zero if `tp+fp < 0` behaviour.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            result = await metric.a_measure(test_case)
+
+            assert round(result, 4) == round(expected_score, 4)
+            assert metric.success == (expected_success)
```
```suggestion
            assert metric.success == expected_success
```

Unless there's a reason for this that I'm not aware of.

exfalsoquodlibet replied:
Good catch. It does not affect results since
```
(True)
>>> True
```
but sloppy coding.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+
+            mock_native_model.a_generate = AsyncMock(
+                return_value=(fact_classification_result, expected_cost)
+            )
```
I imagine this is what we should use for all the tests - can probably be set up in the fixture method.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            )
+
+            metric = FactualCorrectnessMetric(model=mock_native_model)
+            metric.using_native_model = True  # ensure we hit native logic
```
We shouldn't need this line as it's internal details and could potentially conceal a bug. We don't expect the get costs without a native model so we'd expect the test to fail to help reveal the issue.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+        assert metric.is_successful() is False
+
+        metric.error = "Some error"
+        assert metric.is_successful() is False
```
It'd be better practice to test these based on the input into metric rather than changing the class state directly. 

I'd suggest you could use threshold as the means to check basic True and False. For the error state I'm a bit confused how we can fall into that so will leave a comment there.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+        @patch.object(
+            FactualCorrectnessTemplate, "classify_facts", return_value="mocked_prompt"
+        )
+        async def test_non_native_model_returns_valid_schema(
```
For this test I think you'd be best to mock the a_generate function on the model to return the FactClassificationResult object

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            metric = FactualCorrectnessMetric(model=mock_non_native_model)  # type: ignore
+            _ = await metric.a_measure(test_case)
+
+            assert metric.evaluation_cost is None  # should remain None
```
This might be better as an explicit other test (apologies if this conflicts with anything I said earlier) where we assert a non native model doesn't set an evaluation cost.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            assert isinstance(metric.confusion_matrix, ClassifiedFacts)
+            assert (
+                metric.confusion_matrix == fact_classification_result.classified_facts
+            )
```
I think we'll want to assert that a_measure still returns a score value - as I think what we're really trying to assert is that, when given a non native model, a_measure still performs correctly.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+                    TypeError("schema parse failed"),
+                    fixable_json,
+                ]
+            )
```
This looks good 👍

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            )
+
+            metric = FactualCorrectnessMetric(model=mock_non_native_model)  # type: ignore
+            _ = await metric.a_measure(test_case)
```
Like above, it'd be good to have a score assertion too about what this method returns, this could be a good place for it.

kevindew reviewed (commented):
Amazing job @exfalsoquodlibet, I'm really impressed how quickly you're picking up these concepts. I've added a few comments where we're still using aspects of private state or mocking and suggested how we can avoid it. 

Despite the number of comments though this looks really close to being done.

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+            # mock _a_classify_statements to return the sample confusion matrix
+            metric._a_classify_statements = AsyncMock(
+                return_value=sample_confusion_matrix
+            )
```
Yup, your reasoning is correct - though you should also consider as well that the private methods are implementation details of the system under test and ideally you want to have low coupling of the implementation details to the test. One good outcome of avoiding mocking private methods is that you avoid a situation where, because they could be mocked in any other test, that you're either not sure they're tested or need to write unit tests for the private method.

To provide a bit more info:

The ideal is to not need to mock, as that always introduces an element of fictional scenario into the code. But of course you can't always avoid it for side effects, things that are heavy to set-up.

If you do have to mock then you have some options:

Is the method with a side effect called an object that is passed into the class (dependency injection) then you've got something that is an external collaborator outside the unit tested that can be used.

Is there a clear point of something else being called that is already under test, essentially a different unit code? then that is a good candidate for mocking. For example perhaps it's a method that writes to a database from the DB library?

Sometimes though there just isn't any good options and you just have to mock the private method though - we have to do this with JavaScript sometimes for browser interactions that prevent a test from executing: https://github.com/alphagov/govuk-chat/blob/0141f84eddcc0cc917c572898022bd24219c4515/spec/javascripts/modules/conversation-onboarding-flow-spec.js#L135

exfalsoquodlibet replied:
Thanks for this further explanation, Kev!

exfalsoquodlibet commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+        @pytest.mark.asyncio
+        @patch(
+            "govuk_chat_evaluation.rag_answers.custom_deepeval.metrics.factual_correctness.factual_correctness.logging.debug"
+        )
```
Thanks Kev, I have moved to use caplog throughout the tests.

exfalsoquodlibet commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,415 @@
+        ):
+            # fallback case: schema call fails, raw JSON string returned
+            fixable_json = (
+                '{"classified_facts": {"TP": ["fact1"], "FP": ["fact2"], "FN": []}}'
```
Sorry for this, Kev. I reworked this test.

exfalsoquodlibet commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/template.py:
```diff
@@ -0,0 +1,96 @@
+class ClassifiedFacts(BaseModel):
+    TP: list[string]
+    FP: list[string]
+    FN: list[string]
```
Changed, thanks.  It did not seem to affect compliance to the schema but better for consistency.

exfalsoquodlibet commented:
Thanks so much for your thorough re-review, @kevindew. It really helped me better understand and stay grounded in what I’m testing and why. Still lots to learn, but I hope I’ve addressed your comments — keen to take it from here.

kevindew commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py:
```diff
@@ -0,0 +1,148 @@
+            capture_metric_type(self.__name__, async_mode=self.async_mode)
+            return self.score
+        else:
+            self.error = f"Error: no facts were classified. confusion_matrix is empty for input: {input}."
```
What seems perhaps a bit sub-optimal is that we can flag this error saying the confusion matrix was empty when it could also be because we got invalid JSON back (which feels like what we should flag in the error). It also feels a bit weird that we can have a confusion matrix instance that is in effect an error state so we have to be careful with checking it each time before usage.

What I'd suggest doing is that we change `_a_classify_statement` usage so we end up with self.confusion_matrix is ClassifiedFacts or nil, and have the attribute self.confusion_matrix be Optional[ClassifiedFacts]. Then rather than calling a method in the confusion matrix for validity we can just check it's existence (this would also make the __init__ function a bit cleaner if confusion_matrix was nil initially as it's a little weird we set it to a class instance that we then override when we do the work).

Then in `_a_classify_statement` we could change it to set-up the self.error attribute too.

So in psuedo code could be something like:

```
async def _a_classify_statements(
    self, input: str, actual_output: str, expected_output: str
):
    classified_facts = None
    prompt = self.evaluation_template.classify_facts(
        answer=actual_output, ground_truth=expected_output
    )
    if self.using_native_model:
        res, cost = await self.model.a_generate(
            prompt, schema=FactClassificationResult
        )
        if isinstance(cost, (int, float)):
            self.evaluation_cost = (self.evaluation_cost or 0.0) + cost
        classified_facts = res.classified_facts  # type: ignore[arg-type]
    else:
        try:
            res = await self.model.a_generate(
                prompt, schema=FactClassificationResult
            )
            classified_facts = res.classified_facts  # type: ignore[arg-type]
        except TypeError:
            try:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                data_model = FactClassificationResult(**data)
                classified_facts = data_model.classified_facts
            except Exception as inner_e:
                message = f"Failed to parse fallback JSON for test input: {input}",
                logging.error(message, exc_info=inner_e)
                self.error = message

    if classified_facts:
        if classified_facts.has_facts()
            self.confusion_matrix = classified_facts
        else
            self.error = f"Error: no facts were classified. confusion_matrix is empty for input: {input}."
```

(aside: I thought we might find it better to actually set self.confusion_matrix in this method rather than returning it given we're already assigning a self.evaluation_cost in here and I think it'd be more idiomatic if it either returns values with no internal state modifications or performs all internal state modifications expected)

kevindew replied:
Ah yeah of course, that is a very well behaved type checker.

I don't think we actually use the confusion_matrix outside the class so it might be that it's better as actually a regular variable rather than an instance variable and can be a typed argument to a method e.g. `def _generate_reason(self, confusion_matrix: ClassifiedFacts)`

exfalsoquodlibet replied:
We do export it as part of the results: https://github.com/alphagov/govuk-chat-evaluation/blob/0cf9caa4985a2c86b01975f3ab87668050136422/govuk_chat_evaluation/rag_answers/evaluate.py#L146C27-L146C45

kevindew commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/schema.py:
```diff
@@ -0,0 +1,14 @@
+    FN: list[str] = []
+
+    def has_facts(self) -> bool:
+        return bool(self.TP or self.FP or self.FN)
```
If we add more to this we might want to think about unit testing it?

exfalsoquodlibet replied:
Good point. I've deliberately chosen not to unit test `has_facts()` as an isolated method because I think it adds little value to test such a simple boolean check directly.

Instead, following the discussion we have had so far about testing (or at least, my understandng of this) I've tested its behaviour through tests that verify how it affects the overall system output. The method is effectively tested in several scenarios:

- `test_non_native_model_empty_confusion_matrix_triggers_error`L Tests that when `has_facts()` returns `False` (empty lists), we get the expected error logging and NaN result
- `test_returns_score` with `([], [], 0.0)` parametrisation - Tests scoring behaviour when no facts are present
- Exception handling tests - Verify that when JSON parsing fails and we fall back to empty `ClassifiedFacts()`, the system behaves correctly

I thought from our discussions that this approach should be preferred because it:
- Tests the actual behaviour and contract rather than implementation details
- Shows how `has_facts()` integrates with the rest of the system in real scenarios  
- Focuses on testing observable outcomes rather than internal method calls

The `has_facts()` method is essentially just `return bool(self.TP or self.FP or self.FN)` - testing this in isolation feels like testing Python's boolean logic rather than our business logic.

But keen to hear your extra thoughts on this.

kevindew replied:
Sure, so in more classical approaches to unit testing this would typically be tested because it's a public class that is imported into another file.

The argument you've made for not testing holds sound if this is treated as a private class, that wasn't exposed outside of it's implementation (so typically within the file it is used in) - I know this all gets a bit murky in Python without explicit exports. But as this is a separate file it's probably, even in Python world, a distinct unit of code.

But yeah strict approaches aside you're right that it's trivial logic and would think it's only really necessary to unit test if we add more to it.

kevindew commented on govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/schema.py:
```diff
@@ -0,0 +1,14 @@
+    FP: list[str] = []
+    FN: list[str] = []
+
+    def has_facts(self) -> bool:
```
I did wonder whether it might be more appropriate to use a validator than a helper method here given we consider this invalid if all these lists are empty? I imagine if we do that we also have the benefit that if there were any other issues they'd also be caught as an error.

kevindew replied:
Yeah, very happy to have a chat about. 

Also would also want it to be handled gracefully,  my reasoning of it being a validation error is based on this being considered an error state with the code currently and thus we do something of a pseudo validation currently by calling the method and setting an error state

https://github.com/alphagov/govuk-chat-evaluation/blob/b74895a63f09290556c3c24576efe23e644808b7/govuk_chat_evaluation/rag_answers/custom_deepeval/metrics/factual_correctness/factual_correctness.py#L80-L83

I wondered if something more explicit like below, separates it from the logic and offers an earlier exit:

```
try:
            self.confusion_matrix = await self._a_classify_statements(
                test_case.input,
                test_case.actual_output,
                test_case.expected_output or "",
            )
            logging.debug(
                f"Confusion matrix for test input: '{test_case.input}': \n{self.confusion_matrix}"
            )
            return self._finalise_evaluation(test_case.input)
except ValidatorError as e:
           self.error = f"Invalid confusion matrix: {e}"
           return float("nan")
```

but I do agree semantically that it depends on what we consider empty facts to mean. I was interpreting it as "the LLM hasn't followed instructions properly"

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,580 @@
+        @pytest.mark.asyncio
+        @patch(
+            "govuk_chat_evaluation.rag_answers.custom_deepeval.metrics.factual_correctness.factual_correctness.initialize_model"
+        )
```
I don't think we need this mock anymore (and that we get a more realistic test without it)

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,580 @@
+                in captured_logs
+            )
+
+            assert math.isnan(result)
```
I think we should also assert here that metric.error is populated with what we expect

kevindew commented on tests/rag_answers/custom_deepeval/metrics/test_factual_correctness.py:
```diff
@@ -0,0 +1,580 @@
+
+            assert metric.evaluation_cost is None
+            assert isinstance(result, float)
+            assert result == metric.score
```
```suggestion
```

I'd remove this as it's not really a factor in the scenario we're testing (and opens a question of why not assert on other instance variables e.g reason)

kevindew reviewed (commented):
Amazing stuff - I've got a couple of minor comments on the tests and then a more substantial comment on the use of the ClassifiedFacts model.

The suggestion I've made on ClassifiedFacts only have minor impacts on behaviour so do feel free to make a case, if you that feel that way, that they aren't worth doing. I wasn't sure myself but did wonder if we might use this as a precedent for future metrics so thought it might be something to improve now before duplicated.

Again, very sorry for how long it has taken me to review this again.
